ksql-plugins/ksql-rocksdb-config-setter/src/main/java/io/confluent/ksql/rocksdb/KsqlBoundedMemoryRocksDBConfig.java

```diff
@@ -0,0 +1,132 @@
+ * "ksql.plugins.rocksdb.total.memory", and also configures
+ * the shared RocksDB thread pool to use "ksql.plugins.rocksdb.num.background.threads" threads.
+ * </p>
+ * See https://docs.confluent.io/5.3.0/streams/developer-guide/memory-mgmt.html#rocksdb.
+ */
+public class KsqlBoundedMemoryRocksDBConfig implements RocksDBConfigSetter {
+
+  private static final String CONFIG_PREFIX = "ksql.plugins.rocksdb.";
+  private static final String TOTAL_OFF_HEAP_MEMORY_CONFIG = CONFIG_PREFIX + "total.memory";
+  private static final String N_BACKGROUND_THREADS_CONFIG =
+      CONFIG_PREFIX + "num.background.threads";
+
+  private static final double INDEX_FILTER_BLOCK_RATIO = 0.1;
+  private static final int N_MEMTABLES = 6;
+  private static final long BLOCK_SIZE = 4096L;
+
+  private static org.rocksdb.Cache cache;
+  private static org.rocksdb.WriteBufferManager writeBufferManager;
+
+  private static AtomicBoolean configured = new AtomicBoolean(false);
+  private static long totalOffHeapMemory;
+  private static long memtableSize;
+  private static long totalMemtableMemory;
+  private static int numBackgroundThreads;
+
+  public static void configure(final Map<String, Object> config) {
+    configure(config, new Options());
+  }
+
+  @VisibleForTesting
+  static void configure(final Map<String, Object> config, final Options options) {
+    if (configured.getAndSet(true)) {
+      throw new IllegalStateException(
+          "KsqlBoundedMemoryRocksDBConfig has already been configured. Cannot re-configure.");
+    }
+
+    try {
+      limitTotalMemory(config);
+      configureNumThreads(config, options);
+    } catch (IllegalArgumentException e) {
+      configured.set(false);
+      throw e;
+    }
+  }
+
+  @VisibleForTesting
+  static void reset() {
+    configured.set(false);
+  }
+
+  private static void limitTotalMemory(final Map<String, Object> config) {
+    if (!config.containsKey(TOTAL_OFF_HEAP_MEMORY_CONFIG)) {
+      throw new IllegalArgumentException(
+          "Missing configuration: " + TOTAL_OFF_HEAP_MEMORY_CONFIG);
+    }
+    totalOffHeapMemory = Long.parseLong((String)config.get(TOTAL_OFF_HEAP_MEMORY_CONFIG));
+
+    memtableSize = totalOffHeapMemory / 2 / N_MEMTABLES;
+    totalMemtableMemory = N_MEMTABLES * memtableSize;
+
+    cache = new org.rocksdb.LRUCache(totalOffHeapMemory, -1, false, INDEX_FILTER_BLOCK_RATIO);
+    writeBufferManager = new org.rocksdb.WriteBufferManager(totalMemtableMemory, cache);
+  }
+
+  private static void configureNumThreads(final Map<String, Object> config, final Options options) {
+    if (!config.containsKey(N_BACKGROUND_THREADS_CONFIG)) {
+      throw new IllegalArgumentException("Missing configuration: " + N_BACKGROUND_THREADS_CONFIG);
+    }
+    numBackgroundThreads = Integer.parseInt((String)config.get(N_BACKGROUND_THREADS_CONFIG));
+
+    // All Options share the same Env, and share a thread pool as a result
+    options.getEnv().setBackgroundThreads(numBackgroundThreads);
```

Comment:
Why not just set this in `setConfig` then it isn't important that it's ultimately a shared resource.

Reply:
`setConfig()` is called each time a new query is run/streams app is started. By configuring the thread pool once at the beginning, we avoid edge cases where the user has changed the value of `N_BACKGROUND_THREADS_CONFIG` between queries.

Reply:
So we're allowing people to change the properties this setter provides via `SET 'x'='y'` style syntax?

What happens if the user changes this config when starting a new query?  Does it change the number of threads?